packages/flutter_tools/lib/src/commands/run.dart

```diff
@@ -715,6 +715,10 @@ class RunCommand extends RunCommandBase {
         throwToolExit('Lost connection to device.');
       }
       rethrow;
+    } finally {
+      // However we exited from the runner, ensure the terminal has line mode
+      // and echo mode enabled before we return the user to the shell.
+      globals.terminal.singleCharMode = false;
```

Comment:
Does the resident runner modify any other modes that should be reset?

Reply:
So this is actually a convenience setter around two modes:

```
    if (value) {
      stdin.echoMode = false;
      stdin.lineMode = false;
    } else {
      stdin.lineMode = true;
      stdin.echoMode = true;
    }
```

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/terminal.dart#L288

I'll take one more pass at the code to check for anything else suspicious, however.